File: array/c/03_1.c
/*
 * moore voting algo
 * two steps: find major candidate, check if thats the majority or not
 * */
#include<stdio.h>
#include<stdlib.h>

int get_majority_candidate(int *arr, int n)
{
    int i;
    int current_element = arr[0], element_vote = 1;

    for (i = 1; i < n; ++i)
    {
        if (arr[i] == current_element)
            element_vote++;
        else
        {
            // now two cases may aries, element_vote can be greater than 0 or can be 0; if it is 0
            // then change the current_element to arr[i];
            if (element_vote == 0)
            {
                 current_element = arr[i];
                 element_vote++;
            }
            else
                element_vote--;
        }
    }
    return current_element;
}

int check_if_major(int *arr, int n, int candidate)
{
    int i, count = 0;
    for (i = 0; i < n; ++i)
        if (arr[i] == candidate)
            count++;
    return (count > n/2)?1:0;
}

int main()
{
    int *arr, n, i;
    scanf("%d", &n);
    arr = (int*) malloc(n*sizeof(int));
    for (i = 0; i < n; ++i)
        scanf("%d", &arr[i]);
    int majority_candidate = get_majority_candidate(arr, n);
    int val = check_if_major(arr, n, majority_candidate);
    if (val)
        printf("\n%d is the majority candidate \n", majority_candidate);
    else
        printf("\nThere are no majority element in the array \n");
    return 0;
}

File: array/python/03_1.py
def getMajorityCandidate(li):
    majority_candidate = li[0]
    candidate_cnt = 1

    for i in range(1, len(li)-1):
        if majority_candidate == li[i]:
            candidate_cnt += 1
        else:
            if candidate_cnt == 0:
                majority_candidate = li[i]
                candidate_cnt = 1
            else:
                candidate_cnt -= 1

    return majority_candidate


def isMaj(li, candidate):
    cnt = 0
    for i in li:
        if i == candidate:
            cnt += 1

    return cnt >= len(li)/2


li = [5, 2, 3, 78, 5, 5, 5, 5, 6, 5, 4, 5, 5, 5]

if isMaj(li, getMajorityCandidate(li)):
    print getMajorityCandidate(li), 'is the majority element'
else:
    print 'no majority element'

# time is 2*n
